transform/combined_filter.go

```diff
@@ -0,0 +1,274 @@
+func (i *EthCombinedIndexer) ProcessBlock(blk *pbeth.Block) {
+	var keys []string
+	for _, trace := range blk.TransactionTraces {
+		for _, key := range callKeys(trace, CP) {
+			keys = append(keys, key)
+		}
+		for _, key := range logKeys(trace, LP) {
+			keys = append(keys, key)
+		}
+	}
+
+	i.BlockIndexer.Add(keys, blk.Number)
+	return
+}
+
+func (f *CombinedFilter) String() string {
+	//	var addresses []string
+	//	var signatures []string
+	//	for _, a := range p.Addresses {
+	//		addresses = append(addresses, a.Pretty())
+	//	}
+	//	for _, s := range p.EventSignatures {
+	//		signatures = append(signatures, s.Pretty())
+	//	}
+	//	return fmt.Sprintf("LogFilter{addrs: %s, evt_sigs: %s}", strings.Join(addresses, ","), strings.Join(signatures, ","))
+	//
+	return "combinedFilter"
+}
+
+func (f *CombinedFilter) matches(trace *pbeth.TransactionTrace) bool {
+	for _, lf := range f.LogFilters {
+		if lf.matches(trace) {
+			return true
+		}
+	}
+	for _, cf := range f.CallToFilters {
+		if cf.matches(trace) {
+			return true
+		}
+	}
+	return false
+}
+
+func (f *CombinedFilter) Transform(readOnlyBlk *bstream.Block, in transform.Input) (transform.Output, error) {
+	ethBlock := readOnlyBlk.ToProtocol().(*pbeth.Block)
+	traces := []*pbeth.TransactionTrace{}
+	for _, trace := range ethBlock.TransactionTraces {
+		if f.matches(trace) {
+			traces = append(traces, trace)
+		}
+	}
+	ethBlock.TransactionTraces = traces
+	return ethBlock, nil
+}
+
+// GetIndexProvider will instantiate a new index conforming to the bstream.BlockIndexProvider interface
+func (f *CombinedFilter) GetIndexProvider() bstream.BlockIndexProvider {
+	if f.indexStore == nil {
+		return nil
+	}
+
+	if len(f.CallToFilters) == 0 && len(f.LogFilters) == 0 {
+		return nil
+	}
+
+	return transform.NewGenericBlockIndexProvider(
+		f.indexStore,
+		CombinedIndexerShortName,
+		f.possibleIndexSizes,
+		getcombinedFilterFunc(f.CallToFilters, f.LogFilters),
+	)
+
+}
+
+func getcombinedFilterFunc(callFilters []*CallToFilter, logFilters []*LogFilter) func(transform.BitmapGetter) []uint64 {
+	return func(getFunc transform.BitmapGetter) (matchingBlocks []uint64) {
+		out := roaring64.NewBitmap()
+		for _, f := range logFilters {
+			fbit := filterBitmap(f, getFunc, LP)
+			out.Or(fbit)
+		}
+		for _, f := range callFilters {
+			fbit := filterBitmap(f, getFunc, CP)
+			out.Or(fbit)
+		}
+		return nilIfEmpty(out.ToArray())
+	}
+}
+
+func logKeys(trace *pbeth.TransactionTrace, prefix string) (out []string) {
+	for _, log := range trace.Receipt.Logs {
+		var evSig []byte
+		if len(log.Topics) != 0 {
+			// @todo(froch, 22022022) parameterize the topics of interest
+			evSig = log.Topics[0]
+		}
+
+		out = append(out, prefix+hex.EncodeToString(log.Address), prefix+hex.EncodeToString(evSig))
+	}
+
+	return
+}
+func callKeys(trace *pbeth.TransactionTrace, prefix string) (out []string) {
+	for _, call := range trace.Calls {
+		out = append(out, prefix+hex.EncodeToString(call.Address))
```

Comment:
Aren't we adding multiple same `call.Address` when within a trace, the same address is called multiple times? Should we avoid that?